plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataRebuilder.java

```diff
@@ -0,0 +1,466 @@
+import org.sonatype.nexus.repository.storage.BucketEntityAdapter;
+import org.sonatype.nexus.repository.storage.Component;
+import org.sonatype.nexus.repository.storage.ComponentDatabase;
+import org.sonatype.nexus.repository.storage.StorageFacet;
+import org.sonatype.nexus.repository.storage.StorageTx;
+import org.sonatype.nexus.repository.view.Content;
+import org.sonatype.nexus.repository.view.payloads.StringPayload;
+import org.sonatype.sisu.goodies.common.ComponentSupport;
+
+import com.google.common.base.Strings;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import com.orientechnologies.orient.core.command.OCommandResultListener;
+import com.orientechnologies.orient.core.db.ODatabaseRecordThreadLocal;
+import com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx;
+import com.orientechnologies.orient.core.id.ORID;
+import com.orientechnologies.orient.core.metadata.schema.OType;
+import com.orientechnologies.orient.core.record.impl.ODocument;
+import com.orientechnologies.orient.core.sql.query.OSQLAsynchQuery;
+import org.w3c.dom.Document;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * Maven 2 repository metadata re-builder.
+ *
+ * @since 3.0
+ */
+@Singleton
+@Named
+public class MetadataRebuilder
+    extends ComponentSupport
+{
+  private final Provider<DatabaseInstance> databaseInstanceProvider;
+
+  private final BucketEntityAdapter bucketEntityAdapter;
+
+  @Inject
+  public MetadataRebuilder(final @Named(ComponentDatabase.NAME) Provider<DatabaseInstance> databaseInstanceProvider,
+                           final BucketEntityAdapter bucketEntityAdapter)
+  {
+    this.databaseInstanceProvider = checkNotNull(databaseInstanceProvider);
+    this.bucketEntityAdapter = checkNotNull(bucketEntityAdapter);
+  }
+
+  /**
+   * Rebuilds/updates Maven metadata.
+   *
+   * @param repository  The repository who's metadata needs rebuild (Maven2 format, Hosted type only).
+   * @param update      if {@code true}, updates existing metadata, otherwise overwrites them with newly generated
+   *                    ones.
+   * @param groupId     scope the work to given groupId.
+   * @param artifactId  scope the work to given artifactId (groupId must be given).
+   * @param baseVersion scope the work to given baseVersion (groupId and artifactId must ge given).
+   */
+  public void rebuild(final Repository repository,
+                      final boolean update,
+                      @Nullable final String groupId,
+                      @Nullable final String artifactId,
+                      @Nullable final String baseVersion)
+  {
+    checkNotNull(repository);
```

Comment:
Stylistic - these methods' structure is hard to see. It's a bit like:
    step1() {
      // do lots of work
      step2();
    }

Feels like the flow could be clearer by factoring out some methods:

    sql = buildSql(groupId, artifactId, baseVersion)
    step2( sql );